chain/client/src/client.rs

```diff
@@ -1399,26 +1399,39 @@ impl Client {
     }
 
     /// Called asynchronously when the ShardsManager finishes processing a chunk.
+    ///
+    /// Returns an error if the persisting of the completed chunk fails.
     pub fn on_chunk_completed(
         &mut self,
         partial_chunk: PartialEncodedChunk,
         shard_chunk: Option<ShardChunk>,
         apply_chunks_done_callback: DoneApplyChunkCallback,
     ) {
         let chunk_header = partial_chunk.cloned_header();
-        self.chain.blocks_delay_tracker.mark_chunk_completed(&chunk_header);
-        self.block_production_info
-            .record_chunk_collected(partial_chunk.height_created(), partial_chunk.shard_id());
+        let block_height = partial_chunk.height_created();
+        let shard_id = partial_chunk.shard_id();
+
+        if let Err(err) = persist_chunk(partial_chunk, shard_chunk, self.chain.mut_chain_store()) {
+            error!(target: "client", error=?err, chunk_hash=?chunk_header.chunk_hash(),
+                "Failed to persist completed chunk.");
+            if cfg!(debug_assertions) {
+                panic!(
+                    "Failed to persist completed chunk. chunk_hash={:?}, error={:?}",
+                    chunk_header.chunk_hash(),
+                    err
+                );
+            }
```

Comment:
we have [log_assert_fail](https://github.com/near/nearcore/blob/678bce54d772b37f9d4bd604f56fe919e38e6d68/core/o11y/src/lib.rs#L109) exactly for such cases, please take look